### DISCUSSION TITLE
Implicitly support (or ignore) nox

### DISCUSSION THREAD

**Discussion by brettcannon** [microsoft/pylance-release#4089] [Enhancement]
[Nox](https://nox.thea.codes/) is a task runner where you write your tasks in Python code in a `noxfile.py` file. You use the `nox` command which will read your `noxfile.py` and execute it appropriately to run a task.

The issue is that `nox` is a CLI and not a library you install. And since you're writing pure Python code, Pylance typically flags the `import nox` line as not be resolvable. Since most people install `nox` as a CLI tool (or via `pipx run`), you almost never have it somewhere to directly import.

So this feature request is to either implicitly support `import nox` or to ignore the import failure.

**Comment by erictraut** [comment]
I presume that `nox` uses the installed Python interpreter — and that it doesn't include its own interpreter embedded within the CLI. If so, how does the Python runtime resolve the `import nox` statement if `nox` isn't installed as a library? Is there a dynamic import hook that handles the import resolution?

As you are probably aware, pyright doesn't (and cannot) support dynamic import hooks because they involve running arbitrary Python code.

It's unlikely that we'd special-case the `nox` library in pyright, which intentionally has no knowledge of any third-party libraries or import targets.

You probably already know this as well, but you can suppress the import failure diagnostic by adding a `# type: ignore` or `# pyright: ignore` comment after the import statement.

**Reply by brettcannon** [reply]
> I presume that `nox` uses the installed Python interpreter — and that it doesn't include its own interpreter embedded within the CLI.

It uses _an_ interpreter (whichever one it was installed with/for), but you're correct it doesn't ship one in-box.



> If so, how does the Python runtime resolve the `import nox` statement if `nox` isn't installed as a library?

Nox gets installed as an entry point, so it's available to the running interpreter used to run Nox like any other entry point. But the key point is it's very likely not the environment you're developing under where Nox got installed, but some other one dedicated _to_ Nox. For instance, if you used your OS to install Nox or `pipx install`, then Nox will have some virtual enviroment created somewhere on disk just for its use and end up with `nox` on `PATH`. The installation is like any other Python CLI tool, but the fact that you write Python code to configure it is a bit unique.



> You probably already know this as well, but you can suppress the import failure diagnostic by adding a `# type: ignore` or `# pyright: ignore` comment after the import statement.

Yep, and that's what I do, but since I do this every time I create a `noxfile.py` (which is every project I participate in), I noticed a pattern and figured I would open a feature request.
